rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbDeviceStateNode.java

```diff
@@ -0,0 +1,102 @@
+import org.thingsboard.rule.engine.api.TbNodeConfiguration;
+import org.thingsboard.rule.engine.api.TbNodeException;
+import org.thingsboard.rule.engine.api.util.TbNodeUtils;
+import org.thingsboard.server.common.data.DataConstants;
+import org.thingsboard.server.common.data.EntityType;
+import org.thingsboard.server.common.data.id.DeviceId;
+import org.thingsboard.server.common.data.plugin.ComponentType;
+import org.thingsboard.server.common.msg.TbMsg;
+
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+@Slf4j
+@RuleNode(
+        type = ComponentType.ACTION,
+        name = "device state",
+        nodeDescription = "Triggers device connectivity events",
+        nodeDetails = "If incoming message originator is a device," +
+                " sends configured connectivity event for that device to the Rule Engine. " +
+                "Incoming message is forwarded using the <code>Success</code> chain," +
+                " unless an unexpected error occurs during message processing" +
+                " then incoming message is sent further using the <code>Failure</code> chain." +
+                "<br>" +
+                "Supported device connectivity events are:" +
+                "<ul>" +
+                "<li>Connect event</li>" +
+                "<li>Disconnect event</li>" +
+                "<li>Activity event</li>" +
+                "<li>Inactivity event</li>" +
+                "</ul>" +
+                "This is particularly useful when device isn't using transports to receive data," +
+                " such as when fetching data from external API or computing new data within the rule chain.",
+        configClazz = TbDeviceStateNodeConfiguration.class,
+        uiResources = {"static/rulenode/rulenode-core-config.js"},
+        configDirective = "tbActionNodeDeviceStateConfig"
+)
+public class TbDeviceStateNode implements TbNode {
+
+    private static final Map<String, ConnectivityAction> EVENT_ACTIONS = Map.of(
+            DataConstants.CONNECT_EVENT,
+            (ctx, msg) -> ctx.getRuleEngineDeviceStateService()
+                    .onDeviceConnect(ctx.getTenantId(), (DeviceId) msg.getOriginator()),
+            DataConstants.ACTIVITY_EVENT,
+            (ctx, msg) -> ctx.getRuleEngineDeviceStateService()
+                    .onDeviceActivity(ctx.getTenantId(), (DeviceId) msg.getOriginator(), System.currentTimeMillis()),
+            DataConstants.INACTIVITY_EVENT,
+            (ctx, msg) -> ctx.getRuleEngineDeviceStateService()
+                    .onDeviceInactivity(ctx.getTenantId(), (DeviceId) msg.getOriginator()),
+            DataConstants.DISCONNECT_EVENT,
+            (ctx, msg) -> ctx.getRuleEngineDeviceStateService()
+                    .onDeviceDisconnect(ctx.getTenantId(), (DeviceId) msg.getOriginator())
+    );
+
+    @FunctionalInterface
+    private interface ConnectivityAction {
+
+        void execute(TbContext ctx, TbMsg msg);
+
+    }
+
+    String event;
```

Comment:
@ashvayka is it ok to leave the event as a package-private field in order to simplify tests? Or should we refactor this and rewrite the tests?

Same question regarding the couple of methods in the device state service. The reason why we are asking this is because we already have the package-private fields and methods there.

Reply:
I vote to create a separate ticket for test updates and update any package-private fields and methods to remain private.

Reply:
I have checked the tests, and I believe the event may be private. You may execute ```node.init(config with your event)``` to replace your logic.

Reply:
@dskarzh please update the tests.